charts/redpanda/templates/statefulset.yaml

```diff
@@ -68,28 +68,58 @@ spec:
         - name: {{ template "redpanda.name" . }}-configurator
           image: {{ .Values.image.repository}}:{{ template "redpanda.tag" . }}
+          # command: ["/bin/bash", "-c"]
           command: ["/bin/sh", "-c"]
           env:
             - name: SERVICE_NAME
               valueFrom:
                 fieldRef:
                   fieldPath: metadata.name
+            - name: EXTERNAL_NODES
+              value: {{ .Values.external.nodes }}
           args:
             - >
               CONFIG=/etc/redpanda/redpanda.yaml;
               cp /tmp/base-config/redpanda.yaml "$CONFIG";
+              NODE_ID=${SERVICE_NAME##*-};
               {{- if (include "redpanda-atleast-22-1-1" . | fromJson).bool }}
               cp /tmp/base-config/bootstrap.yaml /etc/redpanda/.bootstrap.yaml;
               {{- end }}
               {{- if (include "redpanda-atleast-22-3-0" . | fromJson).bool }}
               rpk --config "$CONFIG" config set redpanda.empty_seed_starts_cluster false;
               {{- else }}
-              NODE_ID=${SERVICE_NAME##*-};
               rpk --config "$CONFIG" config set redpanda.node_id $NODE_ID;
               if [ "$NODE_ID" = "0" ]; then
                 rpk --config "$CONFIG" config set redpanda.seed_servers '[]' --format yaml;
               fi;
               {{- end }}
+              rpk redpanda --config "$CONFIG" config set redpanda.advertised_kafka_api[0].name internal ;
+              rpk redpanda --config "$CONFIG" config set redpanda.advertised_kafka_api[0].address {{ $advertiseAddress }} ;
+              rpk redpanda --config "$CONFIG" config set redpanda.advertised_kafka_api[0].port {{ .Values.listeners.kafka.port }} ;
+              {{- $listenerIndex := 1 -}}
+              {{- range $name, $listener := .Values.listeners.kafka.external -}}
+                {{- $enabled := dig "enabled" $values.external.enabled $listener -}}
+                {{- $listenerNodePortEnabled := and $enabled (eq (dig "type" $values.external.type $listener) "NodePort") -}}
+                {{- $advertiseKafkaHost := $advertiseAddress -}}
+                {{- $advertiseKafkaPort := $listener.nodePort -}}
+                {{- if $listenerNodePortEnabled -}}
+                  {{- if $values.external.nodes -}}
+              NODE_INDEX=`expr $NODE_ID + 1`;
+              NODE_ADDRESS=`echo $EXTERNAL_NODES | cut -d ' ' -f $NODE_INDEX`;
+                    {{- if eq $values.external.nodeType "ip" -}}
+              rpk redpanda --config "$CONFIG" config set redpanda.advertised_kafka_api[{{ $listenerIndex }}].address $NODE_ADDRESS;
+                    {{- else -}}
+              rpk redpanda --config "$CONFIG" config set redpanda.advertised_kafka_api[{{ $listenerIndex }}].address $NODE_ADDRESS.{{ $values.external.domain }};
+                    {{- end -}}
+                  {{- else -}}
+                    {{- $advertiseKafkaHost = printf "$(SERVICE_NAME).%s" $values.external.domain -}}
```

Comment:
Could make use of the template value `redpanda.kafka.external.advertise.address`? Otherwise it could be removed from `_helpers.tpl` as I don't think it's used anywhere.

Reply:
Looking at this and a few other named templates, I see they are using environment variables to allow the container to replace the text with the actual value (`SERVICE_NAME` or `HOST_IP`): https://github.com/redpanda-data/helm-charts/blob/main/charts/redpanda/templates/_helpers.tpl#L115-L141

I think this is an issue because that means these named templates will only work within files that spin up containers, but we don't mention that or block their use anywhere else (in fact it probably isn't possible to do so). I know in the past I've mistakenly tried to use these named templates in ConfigMap or other places and found that the environment variable being referenced is just passed through without ever being associated with the expected value.

I think these named templates should be removed to avoid this confusion, or possibly handled differently so that the chance for confusion is much less. I'll create an issue around this and hopefully that can be handled in a separate PR.

Reply:
https://github.com/redpanda-data/helm-charts/issues/223

Reply:
Sounds like a good idea 😊 Thanks for the investigation work 👍